src/node_api.cc

```diff
@@ -155,6 +155,25 @@ class ThreadSafeFunction : public node::AsyncResource {
       if (mode == napi_tsfn_nonblocking) {
         return napi_queue_full;
       }
+
+      // Return `napi_would_deadlock` because waiting on the queue from a JS
+      // thread will block the event loop of that same thread thus preventing it
+      // from removing items from the queue thereby causing deadlock. Deadlock
+      // can also be caused by one JS thread calling the thread-safe function of
+      // another JS thread because if two JS threads call each other's thread-
+      // safe functions blockingly when both their queues are full they will
+      // both deadlock as they wait on each other.
+      //
+      // Thus, we return `napi_would_deadlock` if we find a Node.js event loop.
+      v8::Isolate* isolate = v8::Isolate::GetCurrent();
+      if (isolate != nullptr) {
+        node::Environment* node_env = node::Environment::GetCurrent(isolate);
+        if (node_env != nullptr) {
+          if (node_env->event_loop() != nullptr)
```

Comment:
We have `GetCurrentEventLoop(isolate)` which would give you the loop directly. That being said, if you have a `Environment` instance, you will also always have a event loop here, so the line that this comment is pointing to is currently redundant.

That being said, isn’t the question whether the event loop is the same event loop as `env->node_env()->event_loop()`, not whether an event loop is currently running at all?

Reply:
@addaleax the question is whether a Node.js event loop is running at all, because we also wish to prevent making a blocking call from one event loop thread to another event loop thread, because such a call could result in mutual deadlock:

0. JSThread1 calls JSThread2 and blocks while its queue is full and because JSThread2's queue is also full.
0. JSThread2 calls JSThread1 before it's had a chance to remove an item from its own queue and blocks because JSThread1's queue is also full.

Now both threads are deadlocked.

It's already undesirable that a JS thread calls `napi_call_threadsafe_function()` blockingly, therefore IMO it should be OK if we're liberal about returning `napi_would_deadlock`.